plugins/fabric8-multi-tenant-manager/src/main/java/com/redhat/che/multitenant/tenantdata/TenantDataProvider.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2016-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package com.redhat.che.multitenant.tenantdata;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.LoadingCache;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.commons.subject.Subject;
+
+/**
+ * Provides {@link UserCheTenantData} for a particular user and cheNamespace type in his tenant.
+ *
+ * @author Oleksandr Garagatyi
+ */
+@Singleton
+public class TenantDataProvider {
+
+  private static final int CACHE_TIMEOUT_MINUTES = 10;
+  private static final int CONCURRENT_USERS = 1000;
+
+  private final LoadingCache<TenantDataCacheKey, UserCheTenantData> tenantDataCache;
+
+  @Inject
+  public TenantDataProvider(TenantDataCacheLoader tenantDataCacheLoader) {
+    this.tenantDataCache =
+        CacheBuilder.newBuilder()
+            .maximumSize(CONCURRENT_USERS)
+            .expireAfterWrite(CACHE_TIMEOUT_MINUTES, TimeUnit.MINUTES)
+            .build(tenantDataCacheLoader);
+  }
+
+  /**
+   * Returns OSO namespace information in form of {@link UserCheTenantData}.
+   *
+   * @param subject subject which tenant data are requested
+   * @param namespaceType type of the namespace which tenant data are requested
+   * @throws InfrastructureException when any error occurs or there is no namespace of specified
+   *     type in user's tenant
+   */
+  public UserCheTenantData getUserCheTenantData(Subject subject, String namespaceType)
+      throws InfrastructureException {
+    checkSubject(subject);
+
+    String keycloakToken = subject.getToken();
+    if (keycloakToken == null) {
```

Comment:
Could you check whether this case is covered by unit tests?

Reply:
Added test case